lib/ClangImporter/ImportDecl.cpp

```diff
@@ -3368,18 +3368,48 @@ namespace {
             }
           }
         }
+      }
+
+      HeaderLoc loc(decl->getLocation());
+      if (returnsRetainedAttrIsPresent && returnsUnretainedAttrIsPresent) {
+        Impl.diagnose(loc, diag::both_returns_retained_returns_unretained,
+                      decl);
+      }
 
-        if (returnsRetainedAttrIsPresent && returnsUnretainedAttrIsPresent) {
-          HeaderLoc loc(decl->getLocation());
-          Impl.diagnose(loc, diag::both_returns_retained_returns_unretained,
-                        decl);
+      if ((!returnsRetainedAttrIsPresent) && (!returnsUnretainedAttrIsPresent)) {
+        if (isForeignReferenceType(decl->getReturnType())) {
+          Impl.diagnose(loc, diag::no_returns_retained_returns_unretained,
+                      decl);
         }
       }
 
       return importFunctionDecl(decl, importedName, correctSwiftName,
                                 std::nullopt);
     }
 
+    static bool isForeignReferenceType(const clang::QualType type) {
+      if (!type->isPointerType())
+        return false;
+
+      auto pointeeType =
+          dyn_cast<clang::RecordType>(type->getPointeeType().getCanonicalType());
+      if (pointeeType == nullptr)
+        return false;
+
+      return hasImportAsRefAttr(pointeeType->getDecl());
+    }
+
+    static bool hasImportAsRefAttr(const clang::RecordDecl *decl) {
+      return decl->hasAttrs() && llvm::any_of(decl->getAttrs(), [](auto *attr) {
+              if (auto swiftAttr = dyn_cast<clang::SwiftAttrAttr>(attr))
+                return swiftAttr->getAttribute() == "import_reference" ||
+                        // TODO: Remove this once libSwift hosttools no longer
+                        // requires it.
+                        swiftAttr->getAttribute() == "import_as_ref";
+              return false;
+            });
+    }
```

Comment:
Looks like these functions are copied-and-pasted from elsewhere in ClangImporter. Let's use the existing API instead.